File: GroupAnagrams.java
import java.util.*;

class GroupAnagrams {
    public static void main(String[] args) {
        String[] strs = {"tan", "ant", "nat", "bat", "ate", "tea", "eat"};
        System.out.println(groupAnagrams(strs));
    }

    public static List<List<String>> groupAnagrams(String[] strs) {
        HashMap <String, List<String>> map = new HashMap<>();
        for (String str : strs) {
            String base = base(str);
            System.out.println("base = " + base);
            
            if (!map.containsKey(base)) {
                map.put(base, new ArrayList<>());
            }

            map.get(base).add(str);
        }

        return new ArrayList<>(map.values());

    }

    public static String base(String str) {
        char[] base = str.toCharArray();
        Arrays.sort(base);
        String res = new String(base);
        return res;
            
    }

    
}

File: GroupAnagrams.py
from typing import List
from collections import defaultdict


class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        map = defaultdict(list)

        for i in strs:
            
            if ''.join(sorted(i)) in map:
                map[''.join(sorted(i))].append(i)
            else: map[''.join(sorted(i))].append(i)
        return map.values()


strs = ["eat","tea","tan","ate","nat","bat"]

print(Solution().groupAnagrams(strs))
